miriway: init at unstable-2022-12-18

OPNA2608 opened the pull request:
###### Description of changes

- Packages [Miriway](https://github.com/Miriway/Miriway), an example desktop environment based on the Mir Wayland compositor
- Adds a module for enabling Miriway and providing a system-wide configuration, if desired (similar to WMs like Sway or Awesome, the provided default / system-wide config can be copied to a user's home and adjusted as needed there)

<details>
  <summary>An example of a configured session</summary>

```
ctrl-alt=t:tym
enable-x11=
add-wayland-extensions=all

shell-component=dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY

shell-component=waybar
shell-component=wbg Pictures/miriway-wallpaper

shell-meta=s:synapse
```

\+ some tinkering with `waybar`:

  ![image](https://user-images.githubusercontent.com/23431373/213951385-62fd219f-0b73-4b48-8815-097a14d573b6.png)

</details>

I gave writing a `nixosTest` a shot but didn't manage to get Mir to launch in a test VM. It tended to either hard-crash on software rendering due to KMS / EGL stuff being unavailable, or just seemingly do nothing Wayland-y / easy to test on the dummy drivers. Might give this a shot another time.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

OPNA2608 committed:
miriway: init at unstable-2022-12-18

OPNA2608 commented:
> Cacheable portion of option doc build failed.
> Usually this means that an option attribute that ends up in documentation (eg `default` or `description`) depends on the restricted module arguments `config` or `pkgs`.

Hmm, that's good to know. Guess I'll adjust that part of the module so it doesn't refer to `pkgs` things anymore.

NickCao commented on nixos/modules/programs/miriway.nix:
```diff
@@ -0,0 +1,62 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
```
Avoid these globally scoped `with`, they make codes hard to read.

OPNA2608 replied:
I do agree with you, but I think this might need a treewide fix. That part was copied verbatim from Sway's module for consistency, and most modules I checked unfortunately use a global `with lib`.

OPNA2608 committed:
nixos/miriway: init

NickCao commented:
Shall we also add a test for this. modules are extra hard to review without them.

OPNA2608 commented:
Sorry about all the pushes.

I have a NixOS test now that WorksForMe:tm: on my x86_64-linux machine and on x86_64-linux CI, but aarch64-linux CI seems to struggle with it.

<details>
  <summary>Opening the first Wayland-native terminal via the window manager's keybind is consistently not happening. In the logs I see that the keybind was fired, but I don't see the startup logs of the terminal itself, and it times out waiting for the OCR to see the terminal's prompt.</summary>

```
(finished: sending monitor command: sendkey ctrl-alt-t, in 0.00 seconds)
machine: waiting for alice@machine to appear on screen
# ...
error: building of '/nix/store/48njg2m0gkydd3phfh339a333vljhg8n-vm-test-run-miriway.drv!out' from .drv file timed out after 3600 seconds
```
</details>

<details>
  <summary>I've made it so it opens the terminal on startup, in case it was trying to fire the keybind before Miriway can process it. That works, and it tries to run a first test command to check the Wayland environment. The command should generate a file with the logs, but it looks like it's stuck waiting for that file now.</summary>

```
(finished: waiting for alice@machine to appear on screen, in 12.93 seconds)
machine: sending keys 'test-wayland
'
# ...
(finished: sending keys 'test-wayland
', in 0.13 seconds)
machine: waiting for file '/tmp/test-wayland-exit-ok'
# ...
cleanup
kill machine (pid 8)
machine # qemu-system-aarch64: terminating on signal 15 from pid 6 (/nix/store/il2zyns5nlpi53sczvgg3mdpkm5z524z-python3-3.10.9/bin/python3.10)
(finished: cleanup, in 0.05 seconds)
Traceback (most recent call last):
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 112, in main
    driver.run_tests()
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
    self.test_script()
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 14, in <module>
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 700, in wait_for_file
    retry(check_file)
  File "/nix/store/vigv5wxf1y4qkclwg1hrjivl7br0w5hs-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 131, in retry
    raise Exception(f"action timed out after {timeout} seconds")
Exception: action timed out after 900 seconds
```
</details>

<details>
  <summary>It looks like it fails to receive the keyboard inputs in general? On x86_64-linux I see that Mir is recognising & registering new evdev input devices when the first keyboard inputs are sent. I checked that it is indeed connected to the inputs getting sent and not a timing thing by making it sleep 2, 3, 5 seconds before sending the keybind.<br/>
On aarch64-linux CI, no such output is sent. Maybe this is a problem with how the VM inputs are configured? In which case I don't think I'm NixOS test / VM savvy enough to know how to address this.</summary>

```
machine # [   22.696751] xsession[870]: [2023-02-03 19:14:40.580481] <information> evdev-input: Added /dev/input/event1: QEMU USB Tablet
machine # [   22.698601] xsession[870]: [2023-02-03 19:14:40.580709] <information> evdev-input: Added /dev/input/event3
machine # [   22.709933] xsession[870]: [2023-02-03 19:14:40.597750] <information> evdev-input: Added /dev/input/event0
machine # [   22.723355] xsession[870]: [2023-02-03 19:14:40.611172] <information> evdev-input: Added /dev/input/event5
machine # [   22.724655] xsession[870]: [2023-02-03 19:14:40.612472] <information> evdev-input: Added /dev/input/event4
```
</details>

I'll check if I can replicate this on my aarch64-linux hardware and try to debug & fix it there, though I'm not sure *how* exactly :sweat:.

OPNA2608 commented:
I've checked by running the test interactively on my aarch64-linux hardware: On the tty keyboard works fine, the moment Miriway is launched it stops being accepted. Not sure what's going wrong there or how to fix it, but since it still works fine on x86_64-linux *and* natively on my aarch64-linux hardware, I've just marked the test broken on ARM for now.

It's not ideal, but I daily-drive Miriway on my ARM laptop so it should hopefully be fine?

NickCao commented on pkgs/applications/window-managers/miriway/default.nix:
```diff
@@ -0,0 +1,78 @@
+
+  postPatch = ''
+    substituteInPlace CMakeLists.txt \
+      --replace "\''${CMAKE_INSTALL_PREFIX}/bin" "\''${CMAKE_INSTALL_BINDIR}" \
+      --replace "/usr/share" "\''${CMAKE_INSTALL_DATADIR}" \
+      --replace "/etc" "\''${CMAKE_INSTALL_SYSCONFDIR}"
```
`CMAKE_INSTALL_FULL_<dir>` actually, see https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html

OPNA2608 replied:
I disagree, the `FULL` variants doesn't offer a single benefit here.

This replaces paths that are used in `install` invocations. `install` is fully `CMAKE_INSTALL_PREFIX`-aware, so there is nothing to be gained from using the `FULL` variables. See [clang](https://github.com/llvm/llvm-project/blob/e8231eaa154396c5ff3c29314cdca8f2e0118200/clang/CMakeLists.txt#L360-L376) as a notably large CMake user where this works perfectly fine for us (except where we have to [*add* `CMAKE_INSTALL_<dir>` usages that upstream missed with a patch](https://github.com/NixOS/nixpkgs/blob/a100acd7bbf105915b0004427802286c37738fef/pkgs/development/compilers/llvm/8/clang/gnu-install-dirs.patch)), and [CMake's `install` docs](https://cmake.org/cmake/help/latest/command/install.html#installing-files) using regular `CMAKE_INSTALL_<dir>` in their examples.

Explicitly using the `FULL` variants really only matters for i.e. pkg-config file generations where a full path is expected, which this package doesn't do.

NickCao replied:
Thanks for giving the context, this make sense now.

NickCao commented on pkgs/applications/window-managers/miriway/default.nix:
```diff
@@ -0,0 +1,78 @@
+  passthru = {
+    updateScript = unstableGitUpdater { };
+    providedSessions = [ "miriway" ];
+    tests.vm = nixosTests.miriway;
```
This seems to be the convention.
```suggestion
    inherit (nixosTests) miriway;
```

OPNA2608 replied:
Agree, will change.

NickCao replied:
I forgot the surrounding `tests = {}` in the initial comment.

OPNA2608 replied:
Whoops, should be correct now.

NickCao reviewed (commented):
Looks solid, two nitpicks.

OPNA2608 committed:
nixos/miriway: add test

NickCao reviewed (approved):
Time for ofborg. (It is a bit faster now than previous days)
